proto/src/op/message.rs

```diff
@@ -629,192 +628,47 @@ impl Message {
         Ok(buffer)
     }
 
-    /// Sign the message, i.e. add a SIG0 record to this Message.
+    /// Finalize the message prior to sending.
     ///
     /// Subsequent to calling this, the Message should not change.
-    #[cfg(feature = "openssl")]
-    pub fn sign<S: MessageSigner>(&mut self, signer: &S, inception_time: u32) -> ProtoResult<()> {
-        debug!("signing message: {:?}", self);
-        let key_tag: u16 = try!(signer.calculate_key_tag());
-
-        // this is based on RFCs 2535, 2931 and 3007
-
-        // 'For all SIG(0) RRs, the owner name, class, TTL, and original TTL, are
-        //  meaningless.' - 2931
-        let mut sig0 = Record::new();
-
-        // The TTL fields SHOULD be zero
-        sig0.set_ttl(0);
-
-        // The CLASS field SHOULD be ANY
-        sig0.set_dns_class(DNSClass::ANY);
-
-        // The owner name SHOULD be root (a single zero octet).
-        sig0.set_name(Name::root());
-        let num_labels = sig0.name().num_labels();
-
-        let expiration_time: u32 = inception_time + (5 * 60); // +5 minutes in seconds
-
-        sig0.set_rr_type(RecordType::SIG);
-        let pre_sig0 = SIG::new(
-            // type covered in SIG(0) is 0 which is what makes this SIG0 vs a standard SIG
-            RecordType::NULL,
-            signer.algorithm(),
-            num_labels,
-            // see above, original_ttl is meaningless, The TTL fields SHOULD be zero
-            0,
-            // recommended time is +5 minutes from now, to prevent timing attacks, 2 is probably good
-            expiration_time,
-            // current time, this should be UTC
-            // unsigned numbers of seconds since the start of 1 January 1970, GMT
-            inception_time,
-            key_tag,
-            // can probably get rid of this clone if the owndership is correct
-            signer.signer_name().clone(),
-            Vec::new(),
-        );
-        let signature: Vec<u8> = try!(signer.sign_message(self, &pre_sig0));
-        sig0.set_rdata(RData::SIG(pre_sig0.set_sig(signature)));
-
-        debug!("sig0: {:?}", sig0);
-
-        self.add_sig0(sig0);
-        Ok(())
-    }
+    pub fn finalize<MF: MessageFinalizer>(
+        &mut self,
+        finalizer: &MF,
+        inception_time: u32,
+    ) -> ProtoResult<()> {
+        debug!("finalizing message: {:?}", self);
+        let finals: Vec<Record> = finalizer.finalize_message(self, inception_time)?;
```

Comment:
I am currently restricting access to modify the `message` (i.e. this is not a mutable borrow of self to finalize_message). The only thing the callee is capable of doing is returning additional records to append.

This may be too specific to the way SIG0 works, but we can revisit in the future if other implementations become necessary.